compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1236,6 +1236,37 @@ impl<'a> Parser<'a> {
 
     /// Parse an indexing expression `expr[...]`.
     fn parse_index_expr(&mut self, lo: Span, base: P<Expr>) -> PResult<'a, P<Expr>> {
+        if self.prev_token.is_ident()
+            && matches!(self.token.kind, token::OpenDelim(Delimiter::Bracket))
+            && self.look_ahead(1, |t| t.is_lit())
+            && self.look_ahead(2, |t| matches!(t.kind, token::Semi))
+            && self.look_ahead(3, |t| t.is_lit())
+            && self.look_ahead(4, |t| matches!(t.kind, token::CloseDelim(Delimiter::Bracket)))
+        {
+            if let token::Ident(symbol, _) = self.prev_token.kind {
+                let ident = symbol.as_str().to_owned();
+                self.bump(); // [
+                if let token::Literal(lit) = self.token.kind {
+                    let lit1 = lit.symbol.as_str().to_owned();
+                    self.bump(); // lit
+                    self.bump(); // ;
+                    if let token::Literal(lit) = self.token.kind {
+                        let lit2 = lit.symbol.as_str().to_owned();
+                        let mut err = self.struct_span_err(
+                            self.token.span,
+                            "expected one of `.`, `?`, `]`, or an operator, found `;`",
```

Comment:
I wonder if there is a better way than hard-coding this list of expected values. I have only skimmed your changes, so I don't know if this is necessary here or not.

Can't we just display the list of expected tokens that is automatically accumulated in the parser?

Reply:
the problem with attach the note to the diagnostic is many token already eaten